### PR TITLE
REF: implement indexes.extension to share delegation

### PR DESCRIPTION
**Pull request by jbrockmendel**
In following steps, I plan to

- remove the DatetimeDelegateMixin entirely and just use inherit_names
- de-duplicate the slightly-different comparison method code from CategoricalIndex vs DatetimelikeIndex
- move the remaining wrapping utilities from datetimelike to extension.py, after double-checking if we can remove any layers
- delegate more methods, some of which depends on smoothing out small idiosyncrasies, e.g. #30627.

### CODE REVIEW

File: pandas/core/indexes/extension.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+"""
+Shared methods for Index subclasses backed by ExtensionArray.
+"""
+from pandas.util._decorators import cache_readonly
+
+
+def inherit_from_data(name, delegate, cache=False):
```
Comment (WillAyd) [inline pandas/core/indexes/extension.py]: Can you annotate?

File: pandas/core/indexes/extension.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+            def fset(self, value):
+                setattr(self._data, name, value)
+
+            fget.__name__ = name
+            fget.__doc__ = attr.__doc__
+
+            method = property(fget, fset)
+
+    elif not callable(attr):
```
Comment (WillAyd) [inline pandas/core/indexes/extension.py]: Hmm somewhat confused by this function. Documented as working on `methods` but seems to accept attributes as well?

Comment (jbrockmendel) [reply pandas/core/indexes/extension.py]: Will update docstring.  This clause is for regular class attributes, in particular a few List[str] that we share

File: pandas/core/indexes/datetimes.py [hunk 3]
```diff
@@ -87,6 +89,17 @@ class DatetimeDelegateMixin(DatetimelikeDelegateMixin):
...
 
 
+@inherit_names(
+    [
+        "_bool_ops",
+        "_object_ops",
+        "_field_ops",
+        "_datetimelike_ops",
+        "_datetimelike_methods",
+    ],
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: you have the cache=True ones first in datetimelike, can you conform

File: pandas/core/indexes/extension.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        def method(self, *args, **kwargs):
+            result = attr(self._data, *args, **kwargs)
+            return result
+
+        method.__name__ = name
+        method.__doc__ = attr.__doc__
+    return method
+
+
+def inherit_names(names: List[str], delegate, cache: bool = False):
```
Comment (jreback) [inline pandas/core/indexes/extension.py]: you could type this -> Callable[[Type[T]], Type[T]] i think

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. very minor comments; does this properly preserve doc-strings? i guess no easy way to test this

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
k merge on green.

**Comment by jbrockmendel** [comment]
green

**Comment by jreback** [comment]
thanks

**Pull request merged**